Fix racy test issues

hanzei opened the pull request:
#### Summary
This PR fixes the racy tests that fail on `master`. There are two groups of fixes in the PR:
First, whenever a `request.CTX` gets passed to another go routine, it needs to get `Clone()`d first. Otherwise, modifying the original one also modifies the child. See https://github.com/mattermost/mattermost/pull/24897 for more details.

Secondly, https://github.com/mattermost/mattermost/pull/24046 introduced race code around starting and stopping the `BatchMigrationWorker`. Starting and stopping the worker in quick succession could cause a race for `worker.stop`. This is fixed by checking if the worker successfully started before starting it and vice versa.

Please note that there is still racy code in the EE repo. This does not cause a broken CI currently, as the EE tests are run without the `-race` flags. I will open a separate thread for that discussion.


#### Ticket Link
NONE

#### Release Note
```release-note
NONE
```

hanzei commented:
/update-branch

hanzei commented on server/channels/app/session.go:
```diff
@@ -115,6 +115,7 @@ func (a *App) GetSession(token string) (*model.Session, *model.AppError) {
 			// gets called from (*WebConn).isMemberOfTeam and revoking a session involves
 			// clearing the webconn cache, which needs the hub again.
 			a.Srv().Go(func() {
+				// TODO(hanzei): Create a clone of c
```
Leaving this TODO here for now.

`RevokeSessionById` accepts a `*request.Context`, but `Clone()` returns a `request.CTX`. I didn't want to go down the rabbit hole of changing the pointer to an interface as part of this PR. WIll do that once https://github.com/mattermost/mattermost/pull/24877 is merged.

lieut-data commented on server/channels/jobs/batch_migration_worker.go:
```diff
@@ -64,7 +66,9 @@ func (worker *BatchMigrationWorker) Run() {
 	// We have to re-assign the stop channel again, because
 	// it might happen that the job was restarted due to a config change.
-	worker.stop = make(chan bool, 1)
+	if worker.closed.CompareAndSwap(true, false) {
```
I'm wondering if we should put this at the top of the method and simply `return` if we fail to `CompareAndSwap`? The logic at present appears to allow `Run()` twice, duplicating the worker even though we guard `worker.stop`. I suspect there's also a cleaner way to approach all of this, but happy to defer so long as this is "correct".

hanzei replied:
There are three cases to distinguish:
1. A freshly created worker
2. A `Run()`ing worker
3. A `Stop()`ed worker.

The code can't distinguish between 1 and 2 as both have `worker.closed` set to `false`. Moving the check to the top of the method doesn't help.

It's feasible to revert the semantics of `closed` to something like `running`, which means the default value `false` would apply for 1 and 3. But that seems like a large change that also needs to be applied to other similar code pieces. I'm hesitant to do that as part of this PR.

agnivade replied:
@hanzei - Can you paste the race output here for us to understand what exactly was happening? This job worker is very tricky to get right and I'd love to have a unit test to lock in the behavior.

hanzei replied:
@agnivade Sure:
```
WARNING: DATA RACE
Read at 0x00c006274038 by goroutine 252171:
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*BatchMigrationWorker).Stop()
      /mattermost/server/channels/jobs/batch_migration_worker.go:88 +0x72
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*Workers).Stop()
      /mattermost/server/channels/jobs/workers.go:84 +0x29b
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*JobServer).StopWorkers()
      /mattermost/server/channels/jobs/server.go:111 +0x115
  github.com/mattermost/mattermost/server/v8/channels/app.(*Server).Shutdown()
      /mattermost/server/channels/app/server.go:756 +0x1867
  github.com/mattermost/mattermost/server/v8/channels/api4.(*TestHelper).ShutdownApp.func1()
      /mattermost/server/channels/api4/apitestlib.go:360 +0x52
{"timestamp":"2023-11-03 21:19:23.369 Z","level":"debug","msg":"Worker finished","caller":"jobs/base_workers.go:43","worker_name":"PostPersistentNotifications"}

Previous write at 0x00c006274038 by goroutine 252154:
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*BatchMigrationWorker).Run()
      /mattermost/server/channels/jobs/batch_migration_worker.go:67 +0xb9
  github.com/mattermost/mattermost/server/v8/channels/jobs.(*Workers).Start.func1()
      /mattermost/server/channels/jobs/workers.go:52 +0x48

Goroutine 252171 (running) created at:
  github.com/mattermost/mattermost/server/v8/channels/api4.(*TestHelper).ShutdownApp()
      /mattermost/server/channels/api4/apitestlib.go:359 +0xfb
  github.com/mattermost/mattermost/server/v8/channels/api4.(*TestHelper).TearDown()
      /mattermost/server/channels/api4/apitestlib.go:378 +0xb1
  github.com/mattermost/mattermost/server/v8/channels/api4.TestUploadLicenseFile.func8()
      /mattermost/server/channels/api4/license_test.go:57 +0x39
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:476 +0x32
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
{"timestamp":"2023-11-03 21:19:23.369 Z","level":"debug","msg":"Worker stopping","caller":"jobs/base_workers.go:59","worker_name":"InstallNotifyAdmin"}
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47
```

agnivade replied:
Oh I see, the `Stop` is being called from a different goroutine. Thanks.

lieut-data reviewed (changes requested):
Thanks, @hanzei! Just blocking to clarify the `Run()` semantics below.

Also, I haven't followed the history of this context, but wondering if there's an opportunity to bring thread safety without remembering to `c.Clone`? I worry we've introduced something that's going to repeatedly cause us problems, especially when Go's notion of a Context is explicitly thread safe.

agarciamontoro commented on server/channels/app/channel.go:
```diff
@@ -1668,8 +1668,11 @@ func (a *App) AddChannelMember(c request.CTX, userID string, channel *model.Chan
 			return nil, err
 		}
 	} else {
+		crctx := c.Clone()
```
Nit: I must say I'm not a fan of this naming, it took me a while to understand it meant cloned request context.

agarciamontoro reviewed (approved):
This looks good! Thanks for the investigation, @hanzei :tada: I have the same concern Jesse shared before: is there a way to avoid all this cloning so that we don't make this mistake again? Unblocking anyway, I don't think we need to do that *now*.

hanzei commented on server/channels/jobs/batch_migration_worker.go:
```diff
@@ -64,7 +66,9 @@ func (worker *BatchMigrationWorker) Run() {
 	worker.logger.Debug("Worker started")
 	// We have to re-assign the stop channel again, because
 	// it might happen that the job was restarted due to a config change.
-	worker.stop = make(chan bool, 1)
+	if worker.closed.CompareAndSwap(true, false) {
+		worker.stop = make(chan bool, 1)
+	}
```
Other works already use this code pattern, like in https://github.com/mattermost/mattermost/blob/5c05ee5b9db122484649bf01a448a285447a3c89/server/platform/services/searchengine/bleveengine/indexer/indexing_job.go#L101-L103.

hanzei replied:
On another note: I don't think we need the `bool` chan here, given that we never write to the channel anyway. It's safe to make it a `chan struct{}`, right?

lieut-data replied:
Thanks, @hanzei! Do we know if all jobs use this code pattern, or do others need to be updated to match it?

hanzei replied:
With the changes in this PR, all jobs use the same pattern.

hanzei commented:
Heads up that I'm trying to adopt the suggestion regarding thread safety from  @lieut-data.

hanzei commented:
I've updated the PR based on @lieut-data suggestion.

All methods that modify a `request.Context` now first create a shallow copy, then to the modification on the copy and return it. This behaviour is close to how `context.Context.WithValue` works. Their name was also changed to `WithX` which makes it more clear that the caller should use the return value.

While that change has made the code easier to read, it did lead to one issue: In `App.DoLogin`, the freshly created Session gets attached to the request context. Previously, that also affected the caller's context, as it was passed as a pointer. Now, it only affects the local one. I've updated the method to return the session so the caller can attach it.

hanzei commented:
@lieut-data @agarciamontoro Ready for another round of reviews

hanzei commented:
@agnivade I would also like to hear your thoughts about the behaviour changes in `request.Context` once you are back in action.

lieut-data commented on server/channels/api4/channel.go:
```diff
@@ -1460,9 +1459,8 @@ func getChannelMember(c *Context, w http.ResponseWriter, r *http.Request) {
-	ctx := c.AppContext
-	ctx.SetContext(app.WithMaster(ctx.Context()))
-	member, err := c.App.GetChannelMember(ctx, c.Params.ChannelId, c.Params.UserId)
+	c.AppContext = app.RequestContextWithMaster(c.AppContext)
+	member, err := c.App.GetChannelMember(c.AppContext, c.Params.ChannelId, c.Params.UserId)
```
Minor nit -- unrelated to this PR -- but it would be awesome to be able to see this written as something like:
```go
	c.AppContext = c.AppContext.With(app.RequestContextWithMaster)
```

hanzei replied:
That is a fabulous idea that I had to adopt. Added via https://github.com/mattermost/mattermost/pull/24971/commits/68fcb562f5f4b4baed550d3305c1709a19242f6b.

lieut-data commented on server/channels/jobs/migrations/worker.go:
```diff
@@ -100,10 +100,10 @@ func (worker *Worker) DoJob(job *model.Job) {
 	}
 
-	cancelContext := request.EmptyContext(worker.logger)
+	var cancelContext request.CTX = request.EmptyContext(worker.logger)
```
Can we just make `WithContext` return the concrete type instead of `request.CTX`? Since `request.Context` already implements that interface, it can be used anywhere that a `request.CTX` is expected. It also seems not idiomatic to have the "constructor" follow the "accept interfaces, return structs" paradigm, but have `Clone` and the sibling methods return an interface from a concrete type.

hanzei replied:
It's complicated. It's possible to change `Context.WithContext(ctx context.Context) *CTX` to `Context.WithContext(ctx context.Context) *Context`. But that would also require changing the signature method of `CTX.WithContext`. See https://goplay.tools/snippet/jh-xx6H5K47. That is a path that I'm hesitant to go down, as it would tear down the boundary between the interface and the implementation.

lieut-data reviewed (changes requested):
This is a tremendous simplification, @hanzei -- nice work!

Two follow ups:
* See below, but I think we need to keep `-race` for `master` only to cut costs, even though it does occasionally obscure failures like these.
* Comments re: the decision to mix and match when to return `*request.Context` and `request.CTX`. I think there might be an opportunity to simplify further here.

hanzei commented on .github/workflows/server-test-template.yml:
```diff
@@ -40,9 +40,7 @@ jobs:
-          if [[ ${{ github.ref_name }} == 'master' ]]; then
-            export RACE_MODE="-race"
-          fi
+          export RACE_MODE="-race"
```
This is only a temporary change to ensure that my PR _actually_ fixes all racy test issues. I will revert https://github.com/mattermost/mattermost/pull/24971/commits/dd09ddd13738df797c9c1a37628c0934b1cbdc0e before merge.

hanzei replied:
Reverted via https://github.com/mattermost/mattermost/pull/24971/commits/e69a0b9cc5096574331e94950887318c34b0c6ea

agnivade commented:
> @agnivade I would also like to hear your thoughts about the behaviour changes in request.Context once you are back in action.

There are some amazing code quality improvements here @hanzei ! (A perfect candidate for dev meeting show-n-tell :wink:)

agarciamontoro commented on server/public/shared/request/context.go:
```diff
@@ -84,72 +86,89 @@ func (c *Context) UserAgent() string {
 
-func (c *Context) Logger() mlog.LoggerIFace {
-	return c.logger
+func (c *Context) With(f func(ctx CTX) CTX) CTX {
+	return f(c)
 }
```
Should this take care of the cloning as well? As in:
```go
func (c *Context) With(f func(ctx CTX) CTX) CTX {
	return f(c.clone())
}
```
It may be unexpected that all other `WithXYZ` functions do clone the context except for this generic one. Also, if it would clone it, we could actually use it everywhere else, as in:
```go
func (c * Context) WithUserAgent(s string) CTX {
	return c.With(func(c) { c.userAgent = s })
}
```

hanzei replied:
If `With` takes care the cloning, it becomes less useful for any caller of `With`, as that would lead to cloning the context twice: Once within `With` and then as part of the called `WithX` method.

E.g. ` c.AppContext.With(sqlstore.RequestContextWithMaster)` would lead to one `clone` in `With` and one in `RequestContextWithMaster` via `WithContext`.

agarciamontoro reviewed (approved):
This is neat, thanks for the work :tada: I really like the `With` pattern!

I left a suggestion below, but I don't have a super strong opinion on it, so approving already. I'd love to discuss it with you, though :)
